File: 2020/Day12/Haskell/RainRisk-Part2.hs
import System.IO
import Control.Monad

printArray:: (Show a, Eq a) => [a] -> String
printArray (x:xs)
    | null xs = show x
    | otherwise = show x ++ "\n" ++ printArray xs

rotate:: (Int, Int) -> Int -> (Int, Int)
rotate (x, y) amt
    | amt <= 0 = (x, y)
    | otherwise = rotate (-y, x) (amt-90)

solve:: [(Char, Int)] -> (Int, Int) -> (Int, Int) -> Int
solve [] _ (x, y) = abs x + abs y
solve ((ins, amt):xs) w@(wx, wy) p@(px, py) 
    | ins == 'F' = solve xs w (px + wx * amt, py + wy * amt)
    | ins == 'R' = solve xs (rotate w amt) p
    | ins == 'L' = solve xs (rotate w (360-amt)) p
    | ins == 'N' = solve xs (wx + amt, wy) p
    | ins == 'E' = solve xs (wx, wy + amt) p
    | ins == 'S' = solve xs (wx - amt, wy) p
    | ins == 'W' = solve xs (wx, wy - amt) p

main:: IO()
main = do
    handle <- openFile "../input.txt" ReadMode
    contents <- hGetContents handle
    let rows = lines contents
    let instructions = map(\x -> (head x, read (tail x)::Int)) rows
    print (solve instructions (1, 10) (0, 0))
    hClose handle

File: 2020/Day12/Python/RainRisk-Part2.py
import sys
import collections
import itertools
import numpy as np
import math

pos = [0, 0]
waypoint = [1, 10]

instructions = []
for line in open(f"{sys.path[0]}/../input.txt").read().splitlines():
    move = line[:1]
    amount = int(line[1:])
    instructions.append((move, amount))

for ins, amount in instructions:
    if ins == 'F':
        pos[0] += waypoint[0] * amount
        pos[1] += waypoint[1] * amount
    elif ins == "R":
        for _ in range(amount//90):
            x, y = waypoint
            waypoint = [-y, x]
    elif ins == "L":
        for _ in range(amount//90):
            x, y = waypoint
            waypoint = [y, -x]
    elif ins == 'N':    waypoint[0] += amount
    elif ins == 'E':    waypoint[1] += amount
    elif ins == 'S':    waypoint[0] -= amount
    elif ins == 'W':    waypoint[1] -= amount
       
print(abs(pos[0]) + abs(pos[1]))